config/file/parser.go

```diff
@@ -15,23 +18,167 @@ const (
 	extJSON extension = ".json"
 )
 
-type configParser struct {
-	parseFunc func(in []byte, dst interface{}) error
-}
-
-func (p configParser) parse(in []byte, dst interface{}) error {
-	return p.parseFunc(in, dst)
+// configParser exposes a method parse to read bytes as a raw config.
+type configParser interface {
+	// parse parses a raw bytes input as a raw config and stores
+	// the resulting value into dst.
+	parse(in []byte, dst *unmarshaledConfig) error
 }
 
+// newParser returns an appropriate parser according to ext, or a non-nil
+// error if ext is not an expected extension.
 func newParser(ext extension) (configParser, error) {
 	switch ext {
 	case extYML, extYAML:
-		return configParser{parseFunc: yaml.Unmarshal}, nil
+		return yamlParser{}, nil
 	case extJSON:
-		return configParser{parseFunc: json.Unmarshal}, nil
+		return jsonParser{}, nil
 	default:
-		return configParser{}, errors.New("unsupported config format")
+		return nil, errors.New("unsupported config format")
+	}
+}
+
+// yamlParser implements configParser.
+type yamlParser struct{}
+
+// parse decodes a raw yaml input in strict mode (unknown fields disallowed)
+// and stores the resulting value into dst.
+func (p yamlParser) parse(in []byte, dst *unmarshaledConfig) error {
+	decoder := yaml.NewDecoder(bytes.NewReader(in))
+	decoder.KnownFields(true)
+	return p.handleError(decoder.Decode(dst))
+}
+
+// handleError handles a raw yaml decoder.Decode error, filters it,
+// and return the resulting error.
+func (p yamlParser) handleError(err error) error {
+	// yaml.TypeError errors require special handling, other errors
+	// (nil included) can be returned as is.
+	var typeError *yaml.TypeError
+	if !errors.As(err, &typeError) {
+		return err
+	}
+
+	// filter out unwanted errors
+	filtered := &yaml.TypeError{}
+	for _, msg := range typeError.Errors {
+		// With decoder.KnownFields set to true, Decode reports any field
+		// that do not match the destination structure as a non-nil error.
+		// It is a wanted behavior but prevents the usage of custom aliases.
+		// To work around this we allow an exception for that rule with fields
+		// starting with x- (inspired by docker compose api).
+		if p.isCustomFieldError(msg) {
+			continue
```

Comment:
Quick question, as I do not know that much about yaml aliases: 
It seems like `decoder.KnownFields(true)` will interpret alias declarations as errors when decoding the raw yaml. But we discard it via `continue`.
I suppose that when decoding the known field refering to the alias the correct value is set, right ?

This means that `KnownFields(true)` is natively incompatible with aliases and this custom implementation is a work-around this limitation ?